File: Greedy/MinJobScheduling/min.py
class Solution:
    def jobSequencing(self, deadline: list[int], profit: list[int]) -> list[int]:
        n = len(deadline)

        jobs = []

        for i in range(n):
            jobs.append((profit[i], deadline[i]))

        # Highest profit first
        jobs.sort(reverse=True)

        maxDeadline = max(deadline)

        slot = [-1] * (maxDeadline + 1)

        count = 0
        totalProfit = 0

        for p, d in jobs:
            # Find latest available slot
            for t in range(d, 0, -1):
                if slot[t] == -1:
                    slot[t] = 1
                    count += 1
                    totalProfit += p
                    break

        return [count, totalProfit]

File: Greedy/MinJobScheduling/min.cpp
#include <bits/stdc++.h>
using namespace  std;


class Solution {
  public:
    vector<int> parent;

    int find(int x) {
        if (parent[x] == x)
            return x;

        return parent[x] = find(parent[x]);
    }
    vector<int> jobSequencing(vector<int> &deadline, vector<int> &profit) {
        // code here
        int n = deadline.size();

        vector<pair<int, int>> jobs;

        int maxDeadline = 0;

        for (int i = 0; i < n; i++) {
            jobs.push_back({profit[i], deadline[i]});
            maxDeadline = max(maxDeadline, deadline[i]);
        }

        sort(jobs.begin(), jobs.end(),
             [](const pair<int, int>& a, const pair<int, int>& b) {
                 return a.first > b.first;
             });

        // parent[x] = latest available slot <= x
        parent.resize(maxDeadline + 1);

        for (int i = 0; i <= maxDeadline; i++)
            parent[i] = i;

        int count = 0;
        int totalProfit = 0;

        for (auto job : jobs) {
            int p = job.first;
            int d = job.second;

            int slot = find(d);

            if (slot > 0) {
                count++;
                totalProfit += p;

                // Slot is now occupied.
                // Next time find(slot), go to slot-1.
                parent[slot] = find(slot - 1);
            }
        }

        return {count, totalProfit};

    }
};
